### ISSUE TITLE
ESP32 SD SPI mode: invalid commands sent to SD card: CMD13, CMD60 while writing blocks (IDFGH-12168)

### ISSUE THREAD

**Issue by MayaPosch** [espressif/esp-idf#13223]
[sigrok_capture_recording_annotations.txt](https://github.com/espressif/esp-idf/files/14358989/sigrok_capture_recording_annotations.txt)
### Answers checklist.

- [X] I have read the documentation [ESP-IDF Programming Guide](https://docs.espressif.com/projects/esp-idf/en/latest/) and the issue is not addressed there.
- [X] I have updated my IDF branch (master or release) to the latest version and checked that the issue is present there.
- [X] I have searched the issue tracker for a similar issue and not found a similar issue.

### General issue report

I have created a project that is heavily based on the SD card recorder (SPI mode) example from ESP-IDF 4.4.x, only targeting the INMP441 MEMS microphone, and writing raw PCM data to the SD card instead of a fixed size WAV file. The project files can be found on my GitHub: https://github.com/MayaPosch/pcm_recorder

I'm using ESP-IDF 5.1.2, but the same issue occurs with 5.1.1. After flashing the firmware on the ESP32 (WROOM-32), I see that it correctly mounts and identifies the inserted microSD card. This (1 GB microSD) card is inserted into an SD card module that operates off 3.3V, with no level shifter circuitry, only pull-ups on the SPI lines.

When I write a test file right after this ('hello.txt') with some text in it, this works fine, and I can read the newly written file afterwards when inserting the microSD card into a reader connected to the PC. The problem begins when the I2S reading task begins, and it starts writing the 16-bit samples to the card, resulting in the following error:

```
sdmmc_cmd: sdmmc_write_sectors_dma: sdmmc_send_cmd returned 0x109
diskio_sdmmc: sdmmc_write_blocks failed (265)
```

I have reduced the SPI bus speed to 2 MHz and connected a logic analyzer (Saleae Logic 8, 24 MHz) to analyze the SPI bus traffic using the SD SPI mode decoder in Sigrok PulseView. This shows that when the writing commences, the ESP32 sends a CMD24 (write block) command, which completes fine, but after that I see it sending subsequent blocks prefixed first with CMD13 (send status) and then with CMD60, the latter of which the card obviously responds to with R1 0xe5 (illegal command), which the ESP32 repeats a few times until one of these CMD60-prefixed blocks gets a `Data rejected (CRC error)` which results in the previously listed error.

Is this an issue with ESP-IDF 4.x APIs not working (properly) in ESP-IDF 5.x and should I port it to the new APIs, or is something else going on? What concerns me is that the initial block was prefixed with a proper CMD24 and that only then the trouble began. I have attached the annotations export from PulseView's SD SPI decoder for reference.

If any more information is required, please let me know.

**Comment by igrr** [comment]
Hi @MayaPosch, would you be able to attach those PulseView recordings for us to take a look?

Based on how the SDMMC driver is written, I don't think we can accidentally send a completely non-existent command down to the SPI layer. I see two possibilities: either there is something wrong with the setting of SPI mode, so that the data is transmitted to the card with a bit shift; or, there is a signal integrity issue on the board being used.

The way you initialize the card in https://github.com/MayaPosch/pcm_recorder/blob/9b28469fdc94ae267fb21dd22f975371417b1ac1/main/main.c#L105 looks fine to me at the first glance, as well.

We'll try to run your project to see if the same issue can be reproduced, and also check the PulseView recordings if you provide them.

**Comment by igrr** [comment]
I do see the recording attached now, looks like I didn't refresh the page, my bad. Please disregard that question.

**Comment by rrtandler** [comment]
Hi @MayaPosch ,

 I think you are running out of the buffer when calling
`fwrite(samples16, 1, bytes_read, f);`
The length should be bytes_read / 2 or, better, samples_read * sizeof(samples16[0]).

**Comment by MayaPosch** [comment]
@rrtandler You're right, that was indeed wrong. I have fixed it in the linked GitHub repository and re-ran the test.

This time around I do not see the CRC error, and the recording seems to continue as normal. When I look at the PCM file on the SD card afterwards on the PC, it is however still 0 bytes. In the logic analyser's trace I can see that it looks better, except that I still see CMD60 all the time (along with CMD61), which I presume is what is causing the SD card to not write the data, despite valid CMD25 and CMD13 being sent by the MCU mixed in with the noise.

I have attached the new annotations file from the SD SPI mode decoder in PulseView.
[pcm_debug_sigrok_capture_recording_with_fixes_annotations_cmd_repl.txt](https://github.com/espressif/esp-idf/files/14418329/pcm_debug_sigrok_capture_recording_with_fixes_annotations_cmd_repl.txt)

This trace is recorded from the moment that the `i2s_read_task` commences.

**Comment by rrtandler** [comment]
@MayaPosch  - The reason for not seeing data in the file might be a missing fclose(f) before unplugging the card and reading it in PC.

**Comment by rrtandler** [comment]
@MayaPosch  - to improve the data transfer, plese consider following the hint in comment:
`	/* ATTENTION: The print and delay in the read task only for monitoring the data by human,
	 * Normally there shouldn't be any delays to ensure a short polling time,
	 * Otherwise the dma buffer will overflow and lead to the data lost */`

**Comment by MayaPosch** [comment]
@rrtandler Thank you for that reminder. I have removed all debug logging and am also writing the raw 32-bit samples directly to the SD card to minimise delays. It sounds much better now.

As for the odd commands in the logic analyser trace, this might be a bug in the Sigrok PulseView's SD card decoder: https://sigrok.org/bugzilla/show_bug.cgi?id=1365

I had to switch back to the older, stable version of Sigrok due to a libusb issue in the nightly builds, but it would seem that in this older PulseView version there is another bug which combined with the flaws in my own code led me to believe that the ESP32 or ESP-IDF was at fault. My apologies for this. 

Closing this topic now as I think all issues have been resolved.
